### DISCUSSION TITLE
data fetching at  consumes  component  vs top component ?

### DISCUSSION THREAD

**Discussion by vaynevayne** [vercel/next.js#48560] [Help]
### Summary

https://beta.nextjs.org/docs/rendering/server-and-client-components 
As you say
> Instead, we recommend colocating data fetching alongside the component that consumes the data

but React recommend place data fetch at top component, then pass props to child component

First of all, I don't fully agree with react's view of putting arrays at the top level, but in rare cases, for good maintenance, we still need to extract to the top level, which is no way

Does nextjs go against the react concept? Most of the time, the data is just enough for the consumer component to use, which is fine, and we like to package it that way, but in rare cases, the parent component also needs the array, so we have to modify the code to put the data in the top component

**Comment by icyJoseph** [comment]
> but React recommend place data fetch at top component, then pass props to child component

Source? It is not so much that React recommends that, but rather, you can fetch data on a component that manages things such as, loading, success, and errors, and then pass the data to other components which are focused only on actually rendering HTML.

There's this great answer from the Vercel team: https://github.com/vercel/swr/discussions/1408#discussioncomment-1383367, which expands into one way to do such thing in combination with SWR.

Now onto what I think is the issue, you are mixing React Server Components with the ~classic~ Components, that we use today.

In React Server Components, since you are already on the server, by definition, you can fetch the data you need on the component, even `await` the data, and from there render what you need, either by handling all of the logic on the component, or still use components that are fully presentational. Not much has changed, but the phrasing is done in such a way that encourages you to leverage the fact that the component can be rendered on the Server.

There's more to this claim, of course. Normally, if you fetch things yourself, you'd run into:

- data that is read by many components, 
- many components might trigger requests for the same data,
- data updates/changes, should propagate efficiently
- data might have to be cached
- race conditions when the UI requires different data, faster, than the data can be fetched over the network
- etc, etc

One could try and handle this, but it is recommended to use a data-framework, such as SWR or React-Query. However the server side fetch in Next.js' RSC, can handle deduplication, caching, etc... so you can have many components fetching the same data, it'll be handled by them:

> so each route segment can request exactly the data it needs without worrying about duplicate requests. Next.js will read the same value from the fetch cache.

There's a bit more to learn and investigate with Server Components, but at the same time, few things change, it is just that we can do more!

**Comment by vaynevayne** [comment]
https://react.dev/learn/managing-state#sharing-state-between-components

I see what you mean, server fetching data only works with a few interfaces, such as /userInfo, etc., similar to react-query, both have major limitations, such as: Request parameters must not come from user interactions, such as those provided by useState. This is client-side state, and forcing react-query to handle client-side state is counterproductive and adds a lot of complexity

**Reply by icyJoseph** [reply]
Right, but lifting state, is but one tool, for example, co-locating state is its complement, https://kentcdodds.com/blog/state-colocation-will-make-your-react-app-faster#what-is-colocated-state. 

> In a sense, both try to do the same, while lifting state is the ability to "do" something, co-location state is the idea of not overdoing that.

I am really not sure about what you mean, who is suggesting to put "client-side state" on react-query? What do you mean with "client-side state"? Perhaps you mean something like, a toggle boolean value, or an option selected by a user on a dropdown?

Here, https://github.com/vercel/next.js/discussions/47227#discussioncomment-5342500, I posted an example of a server component driven by client interactivity. Though it uses query/search parameters. Further down I have written, hastily, about https://github.com/dai-shi/wakuwork which could do what you want. 

Let's draw a hardline between RSC, and the current state of things.

In our current state of things, with Next.js, we do isomorphic rendering, which is just a fancy way of saying, we render on the server, and on the client we hydrate, to lay interactivity to the server side rendered HTML string.

There we can use various constructs for interaction and data fetching. Often people make a partition here, or rather a classification, for the types of state that emerge. 

We have state that's local to a component, or at worst a sub-set of the React tree. Techniques such as lifting up state, colocation, context, composition, are usually good enough there. Surely there'd network data, and what not here, but it is not very prevalent.

Then we have apps that are mostly a web version of what used to be desktop apps. Think of the YouTube video editor, or some kind of online tool to edit photos, I am tempted to throw Figma in there as well, but with caveats. For these apps, you end up with a form of global state, which is all done in the browser, there's a fair deal of network/server data too, but not too much, as the main goal of the application is to build things in the browser. For these apps, tools like Valtio, Zustand, Redux, Jotai, Recoil, and what not, fit perfectly well. 

The typical 3rd type is the, mostly driven by the server kind of apps. A dashboard, a logs watcher, etc., these apps mostly just want to mirror, as much as possible, the current state of the server data. We have the typical CRUD operations, where several components show data, while others have the ability to mutate these data. In these situations, we'd use something like SWR or React-Query. 

As I said on my other comment, as we need to read and write more data, and there's a 1 to many, relation on which components read/write this data, you either end up building your own little framework, or you buy into a data-framework. It has trade-offs, it is probably (hardly) a bit more JS than you'd ship with your own solution, and you have to learn the API, but it works very well.

Also not to say that Valtion, Jotain, Recoil, etc, couldn't do this, just look at Redux RTK Query..., but we are just creating a mental model here.  

Not exhaustive but it gives us a stepping stone to think of how we actually use the frameworks we use.

So when a page loads, data is fetched, and used to show some UI, as an initial frame, and then users interact with the page, and changes are made to the data we have, and to the UI we show.

What's Server Components trying to do? Well, it wants to make that first frame happen entirely server side, at component level, without the use of data-fetching functions from Next.js for example, taking advantage of streaming, and reducing the client side JavaScript that needs to take part of the hydration process. 

Furthermore, it wants to really render server side as much as possible, leaving the actual client interactivity handling to the so called Client Components.

The beta stage, current, Next.js interpretation/implementation of these ideas is not really supporting, just yet, the process of having the client call some server function. This experimental framework does, https://github.com/dai-shi/wakuwork, by listening at runtime to an incoming client action, the server can render the UI as needed. This YouTube video explains it better than I can: https://www.youtube.com/watch?v=ciQdQkT0bZk

Not only that, but the RSC proposal also includes the `use` hook, and the `cache` function in React, to fetch stuff in the style, but client side. Though this is still under development, so I won't expand much more there.

So a lot is happening, hopefully Next.js' RSC (appDir), stabilises soon and we can start to more seriously play with it, for now I am trying to stay on the sidelines, but glad to keep on discussing!
